File: HackerEarth/december-circuits-17/december-circuits-17/src/LeftOrRight.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.StringTokenizer;

class LeftOrRight {
	private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	private static StringTokenizer stz;
	private static ArrayList<HashMap<Integer, Integer>> L, R;
	
	public static void main(String[] args) throws IOException {
		int n, q;
		stz = new StringTokenizer(br.readLine());
		n = Integer.parseInt(stz.nextToken());
		q = Integer.parseInt(stz.nextToken());
		stz = new StringTokenizer(br.readLine());
		int[] arr = new int[n];
		for (int i = 0; i < n; i++) {
			arr[i] = Integer.parseInt(stz.nextToken());
		}

		populate(arr);

		while (q-- > 0) {
			int y, z;
			char d;
			stz = new StringTokenizer(br.readLine());
			y = Integer.parseInt(stz.nextToken());
			z = Integer.parseInt(stz.nextToken());
			d = stz.nextToken().charAt(0);

			System.out.println(solve(arr, y, z, d));
		}
	}

	private static void populate(int[] arr) {
		L = new ArrayList<>(arr.length);
		R = new ArrayList<>(arr.length);

		for (int i = 0; i < arr.length; i++) {
			L.add(new HashMap<Integer, Integer>());
			R.add(new HashMap<Integer, Integer>());
		}
		for (int i = 0; i < arr.length; i++) {
			HashMap<Integer, Integer> currentMap = L.get(i);
			int j = i, distance = 0;
			do {
				if (!currentMap.containsKey(arr[j]))
					currentMap.put(arr[j], distance);
				distance++;

				j = (j - 1 + arr.length) % arr.length;
			} while (j != i);
		}

		for (int i = 0; i < arr.length; i++) {
			HashMap<Integer, Integer> currentMap = R.get(i);
			int j = i, distance = 0;
			do {
				if (!currentMap.containsKey(arr[j]))
					currentMap.put(arr[j], distance);
				distance++;

				j = (j + 1 + arr.length) % arr.length;
			} while (j != i);
		}

	}

	private static int solve(int[] arr, int y, int z, char d) {
		
		HashMap<Integer, Integer> currentMap;
		if (d == 'L')
			currentMap = L.get(y);
		else
			currentMap = R.get(y);

		if (currentMap.containsKey(z))
			return currentMap.get(z);
		else
			return -1;
	}
}

File: HackerEarth/december-circuits-17/left-or-right.cpp
#include<bits/stdc++.h>
using namespace std;
vector<unordered_map<int,int>> L,R;

void populate(vector<int> &arr){
	for(int i=0;i<arr.size();i++){
		L.push_back(unordered_map<int,int>());
		R.push_back(unordered_map<int,int>());
	}
	for (int i = 0; i < arr.size(); i++) {
		unordered_map<int, int> &currentMap = L[i];
		int j = i, distance = 0;
		do {
			if (currentMap.find(arr[j]) == currentMap.end())
				currentMap[arr[j]] = distance;
			distance++;

			j = (j - 1 + arr.size()) % arr.size();
		} while (j != i);
	}

	for (int i = 0; i < arr.size(); i++) {
		unordered_map<int, int> &currentMap = R[i];
		int j = i, distance = 0;
		do {
			if (currentMap.find(arr[j]) == currentMap.end())
				currentMap[arr[j]] = distance;
			distance++;

			j = (j + 1 + arr.size()) % arr.size();
		} while (j != i);
	}
}

int solve(vector<int>&arr,int y,int z,char d) {
	if (d == 'L') {
		if (L[y].find(z) != L[y].end())
			return L[y][z];
		else
			return -1;

	} else {
		if (R[y].find(z) != R[y].end())
			return R[y][z];
		else
			return -1;
	}
}
int main(){
	int n,q;
	scanf("%d%d",&n,&q);
	vector<int>arr(n);
	for(int i=0;i<n;i++){
		scanf("%d",&arr[i]);
	}
	populate(arr);
	while(q--) {
		int y,z;
		char c;
		//scanf(" %d%d%c",&y,&z,&c);
		cin>>y>>z>>c;
		printf("%d\n",solve(arr,y,z,c));
	}
	return 0;
}
